### ISSUE TITLE
Field transformation doesn't affect interception of a method that uses that field when redefining a class.

### ISSUE THREAD

**Issue by homedirectory** [raphw/byte-buddy#1307]
I have been trying to redefine a class and transform one of the class fields by modifying its type, and then transform its getter method by modifying its return type, but to no avail. Also it's surpising to me that no one has mentioned this issue yet, and I searched a lot.

Consider this class:
```java
public class Box {
    public Integer size;
    public Integer getSize() { return size; }
}
```

Which I would like to transform into the following:
```java
public class BoxMod {
    public BigDecimal size;
    public BigDecimal getSize() { return size; }
}
```

To achieve this the bytecode of `BoxMod` would have to meet the following requirements:
- (R1) have the field list entry for `size` modified (from `Integer` to `BigDecimal`)
- (R2) have the constant pool contain a `Fieldref` of the following form: `BoxMod.size:Ljava/math/BigDecimal;`
- (R3) have the getter method access the new `Fieldref` and return `BigDecimal` type.

I have experimented with 3 different approaches:
#### `defineField` & `defineMethod`
- (R1) [FAIL] -- `defineField` leads to an **additional entry** in the field list, which leads to an error in `defineMethod`, since there are now 2 fields with the same name.
- (R2) [FAIL] -- the constant pool contains a new `Fieldref`, but its type is still `Integer`.
- (R3) [FAIL] -- exception thrown for `FieldAccessor.ofField()`, since (R1) failed
```java
DynamicType.Builder<?> builder = new ByteBuddy().redefine(Box.class)
        .name("BoxMod")
        .defineField("size", BigDecimal.class, Visibility.PUBLIC)
        // Cannot resolve field for public java.math.BigDecimal BoxMod.getSize() using net.bytebuddy.dynamic.scaffold.FieldLocator$ForClassHierarchy
        .defineMethod("getSize", `BigDecimal.clss,` Visibility.PUBLIC)
        .intercept(FieldAccessor.ofField("size"));
```

#### `transformField` & `defineMethod`
- (R1) [SUCCESS]
- (R2) [FAIL]
- (R3) [FAIL] -- `defineMethod().intercept()` is applied before the field is transformed, therefore it still refers to the old field with `Integer` type.
```java
DynamicType.Builder<?> builder = new ByteBuddy().redefine(Box.class)
        .name("BoxMod")
        .field(named("size"))
        .transform(new Transformer.ForField(
                (instrumentedType, token) ->
                new FieldDescription.Token(
                        token.getName(), token.getModifiers(), 
                        // newFieldType represents BigDecimal
                        newFieldType.asGenericType(), token.getAnnotations())))
        // Cannot set or get value of public java.math.BigDecimal BoxMod.getSize() using public java.lang.Integer BoxMod.size
        .defineMethod("getSize", BigDecimal.class, Visibility.PUBLIC)
        .intercept(FieldAccessor.ofField("size"));
```
#### `transformField` & `transformMethod`
The same as the previous experiment, since we have to use `intercept()` after `method()` before we can get to `transform()`. So it will fail at `intercept()` stage, because the field hasn't been transformed yet.
```java
DynamicType.Builder<?> builder = new ByteBuddy().redefine(Box.class)
        .name("BoxMod")
        .field(named("size"))
        .transform(new Transformer.ForField(
                (instrumentedType, token) ->
                new FieldDescription.Token(
                        token.getName(), token.getModifiers(), 
                        // newFieldType represents BigDecimal
                        newFieldType.asGenericType(), token.getAnnotations())))
        .method(named("getSize"))
        // can this method call be avoided somehow? I want to get to transform() immediately
        .intercept(FieldAccessor.ofBeanProperty()) 
        .transform(new Transformer.ForMethod(
                (instrumentedType, token) -> 
                new MethodDescription.Token(
                        token.getName(), token.getModifiers(), token.getTypeVariableTokens(),
                        newFieldType.asGenericType(), token.getParameterTokens(), token.getExceptionTypes(),
                        token.getAnnotations(), token.getDefaultValue(), token.getReceiverType())));
```

---
To sum up, how to ensure that field transformation is applied before `method(...).intercept(...)` is applied?

**Comment by homedirectory** [comment]
Another idea that I had is to tell `.method(...).intercept(...)` to use a field that has a specific type and I think it might be possible with `FieldAccessor.of(FieldDescription)`. However, I haven't found a way to instantiate `FieldDescription` from an existing field, but with a modified type.

**Comment by raphw** [comment]
Rather define a new class:

```java
new ByteBuddy().subclass(Object.class)
        .name(Box.class.getName())
        .defineField("size", BigDecimal.class, Visibility.PUBLIC)
        .defineMethod("getSize", `BigDecimal.clss,` Visibility.PUBLIC)
        .intercept(FieldAccessor.ofField("size"));
```

and than replace the old class with it. Replacing a field is a tricky affair and it was not part of the original intend of Byte Buddy, unfortunately it is poorly supported.

**Comment by homedirectory** [comment]
Thanks for the quick reply. This makes it clearer for me and explains why I had difficulties with this task. Seems like I will have to switch my approach to subclassing.
